src/VscPreviewExtension/src/extension.ts

```diff
@@ -0,0 +1,270 @@
+    }
+
+    const docUri = Uri.parse(mdUri.query);
+
+    for (let editor of window.visibleTextEditors) {
+        if (editor.document.uri.toString() === docUri.toString()) {
+            return window.showTextDocument(editor.document, editor.viewColumn);
+        }
+    }
+
+    return workspace.openTextDocument(docUri).then(doc => {
+        return window.showTextDocument(doc);
+    });
+}
+
+// Create a child process(DfmRender) by '_spawn' to render a html
+class PreviewCore {
+    private _spawn: child_process.ChildProcess;
+    private _waiting: boolean;
+    public _isFirstTime: boolean;
+
+    constructor(context: ExtensionContext) {
+        let extpath = context.asAbsolutePath('./DfmParse/PreviewCore.exe');
+        this._spawn = child_process.spawn(extpath);
+        this._waiting = false;
+
+        this._spawn.stdout.on('data', function (data) {
+            let tmp = data.toString();
+            let endCharCode = tmp.charCodeAt(tmp.length - 1);
+            if (!multipleRead) {
+                if (endCharCode == ENDCODE) {
+                    previewResult = tmp;
+                    provider.update(documentUri);
+                }
+                else {
+                    // The first one and the result is truncated
+                    previewResult = tmp;
+                    multipleRead = true;
+                }
+            }
+            else {
+                if (endCharCode != ENDCODE) {
+                    // The result is truncated and this is not the last one
+                    previewResult += tmp;
+                }
+                else {
+                    // The result is truncated and this is the last one
+                    previewResult += tmp;
+                    multipleRead = false;
+                    provider.update(documentUri);
+                }
+            }
+        });
+
+        this._spawn.stderr.on('data', function (data) {
+            console.log("error " + data + '\n');
+        });
+
+        this._spawn.on('exit', function (code) {
+            console.log('child process exit with code ' + code);
+        });
+    }
+
+    private sendtext() {
+        let editor = window.activeTextEditor;
+        if (!editor) {
+            return;
+        }
+        let doc = editor.document;
+        let docContent = doc.getText();
+        let fileName = doc.fileName;
+        let rtPath = workspace.rootPath;
+        let filePath;
+        if (!rtPath) {
+            let indexOfFilename = fileName.lastIndexOf('\\');
+            rtPath = fileName.substr(indexOfFilename - 1);
+            filePath = fileName.substring(0, indexOfFilename);
+        }
+        else {
+            let rtpath_length = rtPath.length;
+            filePath = fileName.substr(rtpath_length + 1, fileName.length - rtpath_length);
+        }
+        if (doc.languageId === "markdown") {
+            let numOfRow = doc.lineCount;
+            this._spawn.stdin.write(rtPath + '\n');
```

Comment:
Are they used to communicate between threads? I think it's too disperse for user to read, but acceptable for communication.